File: ServerStatusNotifierModel/src/main/java/com/github/grzegorzekkk/serverstatusnotifier/serverstatusnotifiermodel/SsnJsonMessage.java
package com.github.grzegorzekkk.serverstatusnotifier.serverstatusnotifiermodel;

import com.github.grzegorzekkk.serverstatusnotifier.serverstatusnotifiermodel.json.DateTypeAdapter;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.io.Serializable;
import java.util.Date;
import java.util.UUID;

@Data
@AllArgsConstructor
@NoArgsConstructor
public class SsnJsonMessage<T> implements Serializable {
    private static final Gson GSON = new GsonBuilder().registerTypeAdapter(Date.class, new DateTypeAdapter()).create();

    private MessageType status;
    private T data;
    private Date creationDate = new Date();
    private UUID clientId;

    public enum MessageType {
        CONSOLE_REQUEST,
        CONSOLE_RESPONSE,
        DATA_REQUEST,
        DATA_RESPONSE,
        AUTH_REQUEST,
        AUTHORIZED_RESPONSE,
        UNAUTHORIZED_RESPONSE
    }

    public String toJsonString() {
        return GSON.toJson(this);
    }

    public static <T> SsnJsonMessage fromJsonString(String json, Class<T> clazz) {
        TypeToken token = TypeToken.getParameterized(SsnJsonMessage.class, clazz);
        return (SsnJsonMessage<T>) GSON.fromJson(json, token.getType());
    }
}

File: ServerStatusNotifier/app/src/main/java/com/github/grzegorzekkk/serverstatusnotifier/client/model/SsnJsonMessage.kt
package com.github.grzegorzekkk.serverstatusnotifier.client.model

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

class SsnJsonMessage<T>(statusArg: MessageType = MessageType.AUTH_REQUEST, dataArg: T? = null) {
    var status: MessageType = statusArg
    var data: T? = dataArg

    enum class MessageType {
        AUTH_REQUEST,
        DATA_RESPONSE,
        UNAUTHORIZED_RESPONSE
    }

    fun toJsonString(): String {
        return gson.toJson(this)
    }

    companion object {
        private val gson = Gson()

        fun <T> fromJsonString(json: String, clazz: Class<T>): SsnJsonMessage<T> {
            val token = TypeToken.getParameterized(SsnJsonMessage::class.java, clazz)
            return gson.fromJson(json, token.type) as SsnJsonMessage<T>
        }
    }
}